### PR TITLE
Register `SpecialUnitary` operation with `DefaultMixed` and `NullQubit`

### LINKED ISSUE #3793: [BUG] `SpecialUnitary` is not trainable with autodiff on `"default.mixed"` device
### Expected behavior

The following code should produce a non-vanishing gradient:

<details>
<summary>Code</summary>

```python
dev = qml.device('default.mixed', wires=1)

@qml.qnode(dev, interface="jax")
def circuit(x):
    qml.SpecialUnitary(x, wires=[0])
    return qml.expval(qml.PauliZ(0))

x = jax.numpy.array([0.2, 0.1, 0.6])
jax.grad(circuit)(x) # Should produce Array([-0.68453361, -0.3422668 ,  0.03581418], dtype=float64)
```

</details>

### Actual behavior

The above code exampe produces a vanishing gradient:

<details>
<summary>Code</summary>

```python
dev = qml.device('default.mixed', wires=1)

@qml.qnode(dev, interface="jax")
def circuit(x):
    qml.SpecialUnitary(x, wires=[0])
    return qml.expval(qml.PauliZ(0))

x = jax.numpy.array([0.2, 0.1, 0.6])
```
```pycon
>>> jax.grad(circuit)(x)
Array([0., 0., 0.], dtype=float64)
```

</details>

### Additional information

This is caused by the following pipeline:
- `"default.mixed"` device does not support `SpecialUnitary`, because it is not registered with the device, and other than `"default.qubit"`, having a matrix is not sufficient for an operation to be supported
- `SpecialUnitary` is being decomposed, transferring the trainability of its parameters to `TmpPauliRot` operations, which have parameters 0.
- `TmpPauliRot` again is not supported (which is intended) and also being decomposed. As the `TmpPauliRot` operations have zero parameter, they are removed in this decomposition step, preventing the gradient computation from succeeding.

For `"default.mixed"`, and in fact any device that can deal with operations that provide a matrix, adding `SpecialUnitary` to the supported operations is sufficient.

For other devices, we might run into trouble in the future if a custom device:
- is to be used with automatic differentiation (i.e. not with a gradient transform)
- can not handle operations that provide a matrix generically

### Source code

_No response_

### Tracebacks

_No response_

### System information

```shell
Name: PennyLane
Version: 0.29.0.dev0
Summary: PennyLane is a Python quantum machine learning library by Xanadu Inc.
Home-page: https://github.com/XanaduAI/pennylane
Author: 
Author-email: 
License: Apache License 2.0
Location: /home/david/venvs/dev/lib/python3.10/site-packages
Requires: appdirs, autograd, autoray, cachetools, networkx, numpy, pennylane-lightning, requests, retworkx, scipy, semantic-version, toml
Required-by: PennyLane-Lightning, PennyLane-qiskit

Platform info:           Linux-5.15.0-60-generic-x86_64-with-glibc2.35
Python version:          3.10.6
Numpy version:           1.23.5
Scipy version:           1.9.3
Installed devices:
- default.gaussian (PennyLane-0.29.0.dev0)
- default.mixed (PennyLane-0.29.0.dev0)
- default.qubit (PennyLane-0.29.0.dev0)
- default.qubit.autograd (PennyLane-0.29.0.dev0)
- default.qubit.jax (PennyLane-0.29.0.dev0)
- default.qubit.tf (PennyLane-0.29.0.dev0)
- default.qubit.torch (PennyLane-0.29.0.dev0)
- default.qutrit (PennyLane-0.29.0.dev0)
- null.qubit (PennyLane-0.29.0.dev0)
- lightning.qubit (PennyLane-Lightning-0.28.2)
- qiskit.aer (PennyLane-qiskit-0.28.0)
- qiskit.basicaer (PennyLane-qiskit-0.28.0)
- qiskit.ibmq (PennyLane-qiskit-0.28.0)
- qiskit.ibmq.circuit_runner (PennyLane-qiskit-0.28.0)
- qiskit.ibmq.sampler (PennyLane-qiskit-0.28.0)
```

### Existing GitHub issues

- [X] I have searched existing GitHub issues to make sure the issue does not already exist.

### PR DESCRIPTION
**Pull request by dwierichs**
**Context:**
#3650 adds the `SpecialUnitary` operation, including a decomposition. This enables devices already to apply the new operation. However, this is done by computing the matrix and "decomposing" to a `QubitUnitary`

**Description of the Change:**
To enable direct support, and potentially reverting the decomposition direction (`QubitUnitary->SpecialUnitary` instead of `QubitUnitary->SpecialUnitary`)  later on, `SpecialUnitary` is registered  as available operation in the default devices.

In addition, this PR lints the tests for the `NullQubit` device.

**Benefits:**
Direct support of `SpecialUnitary`, potential future decomposition pipeline allowing for differentiability of matrix ops for any number of wires.

**Possible Drawbacks:**
N/A

**Related GitHub Issues:**
fixes #3793

### CODE REVIEW

**Comment by dwierichs** [comment]
[sc-32294]

**Comment by dwierichs** [comment]
This is very likely not needed in the end.
Forward pass support via `QubitUnitary` is sufficient, differentiation is implemented via a decomposition of `SpecialUnitary`, and decomposing `QubitUnitary` back to `SpecialUnitary` is a whole separate topic. Should we go for such a decomposition, we could reactivate/include this PR in that effort.

**Comment by dwierichs** [comment]
Reopening because this PR fixed #3793.

File: tests/devices/test_null_qubit.py [hunk 1]
```diff
@@ -14,9 +14,8 @@
 """
 Unit tests for the :mod:`pennylane.plugin.NullQubit` device.
 """
-import cmath
```
Comment (dwierichs) [inline tests/devices/test_null_qubit.py:17]: This was an unused import.

File: tests/devices/test_null_qubit.py [hunk 2]
```diff
@@ -14,9 +14,8 @@
...
-
 # pylint: disable=protected-access,cell-var-from-loop
+from collections import defaultdict
```
Comment (dwierichs) [inline tests/devices/test_null_qubit.py:18]: Just changed import order.

File: tests/devices/test_null_qubit.py [hunk 3]
```diff
@@ -25,12 +24,18 @@
...
-from collections import defaultdict
 
+@pytest.fixture(
+    name="nullqubit_device",
+    scope="function",
+    params=[(np.float32, np.complex64), (np.float64, np.complex128)],
+)
+def fixture_nullqubit_device(request):
+    """Create a NullQubit device with the indicated number of wires,
+    parametrizing over real and complex dtypes."""
```
Comment (dwierichs) [inline tests/devices/test_null_qubit.py:35]: Giving the fixture callable a different name and providing the name with which it is supposed to be used (via the `name` kwarg) avoids a Pylint message.

**Review by Qottmann — APPROVED** [review]
Looks good @dwierichs 👍

I saw a small typo maybe worth fixing here in the docstring "beyong" in [special_unitary.py#L225](https://github.com/PennyLaneAI/pennylane/blob/master/pennylane/ops/qubit/special_unitary.py#L225) (really nice docstring by the way)

File: pennylane/ops/qubit/special_unitary.py [hunk 4]
```diff
@@ -418,7 +418,7 @@ def compute_matrix(theta, num_wires):
 
             U(\theta) = \exp(i\sum_{m=1}^d \theta_m P_m)
 
-        See :func:`~.ops.qubit.matrix_ops.pauli_basis_strings` for the ordering of Pauli words.
+        See :func:`~.ops.qubit.special_unitary.pauli_basis_strings` for the ordering of Pauli words.
```
Comment (timmysilv) [inline pennylane/ops/qubit/special_unitary.py:421]: not entirely sure what it takes for Sphinx to include this module-level function, but it doesn't appear to exist on [latest](https://docs.pennylane.ai/en/latest/code/api/pennylane.SpecialUnitary.html) or on [your PR's docs](https://xanaduai-pennylane--3651.com.readthedocs.build/en/3651/code/api/pennylane.SpecialUnitary.html). In other words, the new link is busted because it's pointing to nothing :/ could you make it visible in this PR?

**Review by timmysilv — APPROVED** [review]
pending the docs fix (you can even ignore it if it's not the most important in your opinion!)

File: pennylane/ops/qubit/special_unitary.py (hunk 4 above)
Comment (dwierichs) [reply pennylane/ops/qubit/special_unitary.py:421]: I guess I'd have to add it somewhere manually to the docs? :thinking:

File: tests/devices/test_null_qubit.py [hunk 5]
```diff
@@ -876,7 +869,6 @@ def test_hermitian_two_wires_identity_expectation(self, nullqubit_device, theta,
         assert dev.var(obs) == [0.0]
 
 
-@pytest.mark.parametrize("inverse", [True, False])
```
[the remark this answers is not in the document]
Comment (dwierichs) [reply tests/devices/test_null_qubit.py:879]: Removed it.

File: pennylane/ops/qubit/special_unitary.py (hunk 4 above)
Comment (dwierichs) [reply pennylane/ops/qubit/special_unitary.py:421]: I noticed that the ordering of Pauli words also is discussed in the class docstring, so I refer to that instead of the `pauli_basis_strings` function :)

Comment (dwierichs) [reply pennylane/ops/qubit/special_unitary.py]: Resolving :)
